design/20210203.certificate-request-identity.md

```diff
@@ -0,0 +1,165 @@
+---
+title: Certificate Request Identity
+authors:
+  - "@joshvanl"
+reviewers:
+  - "@joshvanl"
+approvers:
+  - "@joshvanl"
+editor: "@joshvanl"
+creation-date: 2021-02-03
+last-updated: 2021-02-03
+status: provisional
+---
+
+# Identity
+
+## Table of Contents
+
+TODO
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [API Changes](#api-changes)
+  - [Upgrading](#upgrading)
+  - [Test Plan](#test-plan)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Certificate Identity Pass Down](#certificate-identity-pass-down)
+<!-- /toc -->
+
+## Summary
+
+In order to implement policy and improve auditing introspection of certificates
+requested and signed by cert-manager, cert-manager must implement a better
+identity framework. Identity in this context is a user or machine which requests
+a signed certificate from cert-manager. This identity will be managed and
+enforced on CertificateRequest resources by cert-manager.
+
+
+## Motivation
+
+The identity of a requester forms a large part of an automated policy system
+that needs to decide whether a certificate should or shouldn't be signed. By
+creating a trusted source of identity for `CertificateRequest`s, a policy engine
+is able to determine whether that identity is allowed to request that
+certificate, given some policy configuration setup by a cluster administrator.
+
+Although auditing exists in Kubernetes and exposes the identity of the
+requester, its configuration is not always exposed to end users, such as
+when using managed Kubernetes (GKE, EKS etc). In scenarios where
+configuring auditing is available, it is often not preferable and an
+anti-pattern to make runtime decisions on historical audit logs.
+
+The upstream [Kubernetes certificates
+`CertificateSigningRequest`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#certificatesigningrequest-v1-certificates-k8s-io)
+API has identity as part of the resource. As the project intends to align and
+likely make this resource compatible with cert-manager in some way, the
+`CertificateRequest` should expose identical identity details. This means
+extensions or additions to cert-manager (policy, auditing) can have the same
+guarantees about what information is available from both resources.
+
+
+### Goals
+
+- Introduce identity to the `CertificateRequest` resource which is securely
+    enforced by cert-manager
+- Ensure the identity exposed in `CertificateRequest`s has parity with the
+    upstream Kubernetes `CertificateSigningRequest` resource
+- Enable external tooling or additions to cert-manager that can make use of this
+    identity in a useful way
+
+### Non-Goals
```

Comment:
We haven't actually mentioned the fact that cert-manager doesn't support the concept of 'Approval' vs the upstream type here.

This is something we can't quite do today as we don't have the ability to create a custom `approval` subresource on our CRD: https://github.com/kubernetes/kubernetes/blob/7a94debba5f8c21bbf8b42b2a7f1d5e974ddb837/plugin/pkg/admission/certificates/approval/admission.go#L78 (the best we _could_ do is perform this SubjectAccessReview within the `status` update handler, so that we can still gate who is allowed to approve despite the _slightly_ different API call semantics).

We *can* use RBAC to limit who is able to `sign` a resource 1:1 with the upstream type, as it utilises the usual `status` subresource: https://github.com/kubernetes/kubernetes/blob/7a94debba5f8c21bbf8b42b2a7f1d5e974ddb837/plugin/pkg/admission/certificates/signing/admission.go#L79 - I think it could be worth calling this out here (and also maybe summarising _somewhere_ the divergences we have)

Reply:
I think I'm quite keen to leave this out tbh. It diverges a bit from the goal of the design doc, and doesn't really have an impact on adding these fields.

It does make sense to me to add the ability to introduce an "approved" condition, but that is likely a separate design document itself.